homeassistant/components/landisgyr_heat_meter/config_flow.py

```diff
@@ -0,0 +1,136 @@
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_DEVICE
+from homeassistant.exceptions import HomeAssistantError
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_MANUAL_PATH = "Enter Manually"
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_DEVICE): str,
+    }
+)
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Ultraheat Heat Meter."""
+
+    VERSION = 1
+
+    async def async_step_user(self, user_input=None):
+        """Step when setting up serial configuration."""
+        errors = {}
+
+        if user_input is not None:
+            if user_input[CONF_DEVICE] == CONF_MANUAL_PATH:
+                return await self.async_step_setup_serial_manual_path()
+
+            dev_path = await self.hass.async_add_executor_job(
+                get_serial_by_id, user_input[CONF_DEVICE]
+            )
+
+            try:
+                return await self.validate_and_create_entry(dev_path)
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+
+        ports = await self.get_ports()
+
+        schema = vol.Schema({vol.Required(CONF_DEVICE): vol.In(ports)})
+        return self.async_show_form(step_id="user", data_schema=schema, errors=errors)
+
+    async def async_step_setup_serial_manual_path(self, user_input=None):
+        """Set path manually."""
+        errors = {}
+
+        if user_input is not None:
+            dev_path = user_input[CONF_DEVICE]
+            try:
+                return await self.validate_and_create_entry(dev_path)
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+
+        schema = vol.Schema({vol.Required(CONF_DEVICE): str})
+        return self.async_show_form(
+            step_id="setup_serial_manual_path",
+            data_schema=schema,
+            errors=errors,
+        )
+
+    async def validate_and_create_entry(self, dev_path):
+        """Try to connect to the device path and return an entry."""
+        model, device_number = await self.validate_ultraheat(dev_path)
+
+        await self.async_set_unique_id(device_number)
+        self._abort_if_unique_id_configured()
+        data = {
+            CONF_DEVICE: dev_path,
+            "model": model,
+            "device_number": device_number,
+        }
+        return self.async_create_entry(
+            title=model,
+            data=data,
+        )
+
+    async def validate_ultraheat(self, port: str):
+        """Validate the user input allows us to connect."""
+
+        reader = UltraheatReader(port)
+        heat_meter = HeatMeterService(reader)
+        try:
+            async with async_timeout.timeout(10):
+                # validate and retrieve the model and device number for a unique id
+                data = await self.hass.async_add_executor_job(heat_meter.read)
+                _LOGGER.debug("Got data from Ultraheat API: %s", data)
```

Comment:
Please only wrap the line that can raise in the try... except block. We can move the logging down.